### DISCUSSION TITLE
transform_video_subject()

### DISCUSSION THREAD

**Discussion by laynr** [TorchIO-project/torchio#681] [Ideas]
### Idea: 
It would be neat to have a transform_video_subject() function.  Similar to [plot_subject()](https://github.com/fepegar/torchio/blob/d56caf5ab9e23d86ddb0472c3f7c4d47e0116bb1/torchio/visualization.py#L103), where I can provide a torchio.Subject, but instead of just plotting/displaying the middle slice  it generates every slice and creates a video transfor as show in https://torchio.readthedocs.io/auto_examples/plot_video.html#sphx-glr-download-auto-examples-plot-video-py.

### Background: 
My current code to generate a dataset looks like this:
```

def data_preparation(patients):
    subjects  = []
    labels_df = pd.read_csv(data_dir / 'train_labels.csv', index_col=0)
    # loop thru patients
    for patient in patients:
        # get label for patient
        label = labels_df._get_value(int(patient), 'MGMT_value')
        # create subject object for each patient
        subject = tio.Subject(
            BraTS21ID=patient,
            MGMT_value=label,
            FLAIR=tio.ScalarImage(training_dir / patient / 'FLAIR',),
            T1w=tio.ScalarImage(training_dir / patient / 'T1w',),
            T1wCE=tio.ScalarImage(training_dir / patient / 'T1wCE',),
            T2w=tio.ScalarImage(training_dir / patient / 'T2w',),
         )
        # add subject object to subjects list
        subjects.append(subject)

    # preprocessing transforms
    preprocessing_transforms = tio.Compose([
        tio.ToCanonical(),
        tio.Resample(1, image_interpolation='bspline'),
        tio.Resample('T1w', image_interpolation='nearest'),
        tio.RescaleIntensity((-1, 1)),
        tio.CropOrPad((280, 280, 264)),
        tio.OneHot(),
    ])
        

    # create datasets from transformed subjects
    dataset = tio.SubjectsDataset(subjects, transform=preprocessing_transforms)
    print(f'patients :{len(dataset)}')

    
    return dataset

# create training and validation datasets    
dataset = data_preparation(patients) 

```
I then iterate thru the dataset, plot, and save the subjects:
```
    for subject in tqdm(iterable):
        filename = out_dir / f'{subject["BraTS21ID"]}_{subject["MGMT_value"]}.png'
        subject.plot(reorient=False, output_path=filename)
```
Full Kaggle notebook available at: https://www.kaggle.com/ohbewise/s4p1-collage
This creates awesome synchronized plots of the middle slice like this:
![plot](https://user-images.githubusercontent.com/775607/135853998-67b0348c-bed0-42c0-84d1-1b0d872cc583.png)

Today I am going to be working in `matplotlib.pyplot` to save every slice of the subject and create **12 synchronized side-by-side** plot videos similar to this video of one image:

https://user-images.githubusercontent.com/775607/135856153-6eb3ed92-834a-48ba-a244-d9fb62b40776.mp4

I feel it would be beneficial for others so I am creating a new discussion here.

### Other
In my use case I will be also removing all plot titles, subtitles, and x/y scaling - smashing all 12 images together by cropping out the extra pixels, so they can be used in 2D machine learning.

**Comment by laynr** [comment]
Will also be discussed at https://www.kaggle.com/c/rsna-miccai-brain-tumor-radiogenomic-classification/discussion/276371

**Comment by fepegar** [comment]
I think this would be very cool. It could be a new method `plot_animation`, for example. Thanks for opening the discussion!

Have you also seen the new `to_gif` method of `Image`? It might offer some tips for the implementation. Maybe the code could be even reused.

P.S.: good luck with the challenge! I think it's tougher than it looks.

**Comment by laynr** [comment]
@fepegar does this look right?   I don't know MRI's well enough to know if it is a synchronized animations
![movie-60](https://user-images.githubusercontent.com/775607/135940273-273268d8-3539-488a-915d-8a6ac23fe81f.gif)

I not going for most efficient code, but I am going for synchronized MRI accuracy.  
I basically just modified  `plot_volume` to take `indices` as a argument and then called it 'depth of volume' times. Something like below:

```python
for x in range(<depth of volume>):
    subject.plot(indices= (x,x,x), output_path=filename, show=False) 
```

visualization.py

```python
    if indices is None:
        indices = np.array(data.shape) // 2
    i, j, k = indices
    slice_x = rotate(data[i, :, :], radiological=radiological)
    slice_y = rotate(data[:, j, :], radiological=radiological)
    slice_z = rotate(data[:, :, k], radiological=radiological)
```

Then used `imageio` to make gif. Notebook is forthcoming.

**Reply by laynr** [reply]
- Full resolution can be see [here](https://storage.googleapis.com/kagglesdsdata/datasets/1628949/2676982/processed/00688/gif/00688.gif?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=databundle-worker-v2%40kaggle-161607.iam.gserviceaccount.com%2F20211005%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20211005T011714Z&X-Goog-Expires=345599&X-Goog-SignedHeaders=host&X-Goog-Signature=2c42546c6ab1cc2e12e847b7db1d8deb8baca5216a5b99df86c76858b032bcdb863faef30554045432c18c3d065bf2227b368cdb874cb86652d2933138f35fcebcfbe406fea1132e16b9894fd7976c40748832c42406bbe637800245ae50663e8ccdf148785d3c2878524eeb4aea7d389dd2e16a4dc323c64778cfac2b2e2fd5723f484f0182d7dda1eaa433c50c9c299e6572f2a76638b21cbae141329fcd1102166e49d2998f3488af89157347dd86fad6be740290d5f9cf471b13566aeb4ae371495515ebd8ad86e2609d2559e61956b1e84223e71b4eab041fb2470286fc191133558bdceea9efbfab5b28792ba76863117cd07702142264416f5eecf212) the one above had to be resized to comply with gitlab's 10mb upload limit.
 - This is the [notebook ](https://www.kaggle.com/ohbewise/synchronized-mri-animations-flair-t1w-t1wce-t2w) used to create gif.
 - ~~[Pull request](https://github.com/fepegar/torchio/pull/682) summitted, do as you want with it.~~
 - Updated [pull request](https://github.com/fepegar/torchio/pull/683) from branch summitted, do as you want with it.
